### PR TITLE
Signup

### PR DESCRIPTION
**Pull request by FearLeSS-21**

### CODE REVIEW

File: .idea/.gitignore
Comment (reemessam) [inline .idea/.gitignore:1]: add .idea to .gitignore

File: src/main/java/org/example/validation/UserValidator.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
+package org.example.validation;
+
+import org.example.model.User;
+
+public class UserValidator {
```
Comment (reemessam) [inline src/main/java/org/example/validation/UserValidator.java:5]: use springboot validation instead [hint](https://reflectoring.io/bean-validation-with-spring-boot/)

File: src/main/java/org/example/controller/UserController.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    private UserService userService;
+
+    @PostMapping("/signup")
+    public ResponseEntity<?> signup(@RequestBody User user) {
+        String validationError = UserValidator.validateUser(user);
+        if (validationError != null) {
+            return ResponseEntity.badRequest().body(validationError);
+        }
+
+        User existingUser = userService.findByEmail(user.getEmail());
```
Comment (reemessam) [inline src/main/java/org/example/controller/UserController.java:25]: move the logic inside the service.registerUser

File: src/main/java/org/example/repository/UserRepository.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package org.example.repository;
+
+import org.example.model.User;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface UserRepository extends JpaRepository<User, Long> {
+    User findByEmail(String email);
```
Comment (reemessam) [inline src/main/java/org/example/repository/UserRepository.java:9]: make it return optional

File: src/main/java/org/example/service/UserService.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+    @Autowired
+    private UserRepository userRepository;
+
+    @Autowired
+    private PasswordEncoder passwordEncoder;
+
+    // Registers a new user with an encrypted password
+    public User registerUser(User user) {
+        // Encrypt the plain text password before saving to the database
```
Comment (reemessam) [inline src/main/java/org/example/service/UserService.java:20]: check for email here and throw if exists and handle the exception you will also need to handle the exception when you use spring boot validation [hint](https://www.geeksforgeeks.org/exception-handling-in-spring-boot/)

File: src/test/java/org/example/SignupIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    @Autowired
+    private UserRepository userRepository;
+
+    @Autowired
+    private PasswordEncoder passwordEncoder;
+
+
+    @Test
+    public void testSuccessfulSignup() throws Exception {
+        String newUserJson = "{\"email\": \"testuser@example.com\", \"password\": \"Password123!\", \"name\": \"Test User\"}";
```
Comment (reemessam) [inline src/test/java/org/example/SignupIntegrationTest.java:37]: create the user and then use object mapper to map it to json

File: src/test/java/org/example/SignupIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+        mockMvc.perform(MockMvcRequestBuilders.post("/users/signup")
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content(newUserJson))
+                .andExpect(status().isOk())
+                .andDo(print())
+                .andExpect(content().string(containsString("testuser@example.com")))
```
Comment (reemessam) [inline src/test/java/org/example/SignupIntegrationTest.java:44]: assert it was saved to the db

File: src/test/java/org/example/SignupIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+                .andExpect(content().string(containsString("Test User")));
+    }
+
+    @Test
+    public void testSignupWithExistingEmail() throws Exception {
+        // Pre-create a user in the database
+        User existingUser = new User();
```
Comment (reemessam) [inline src/test/java/org/example/SignupIntegrationTest.java:51]: use builder to create the user

File: src/test/java/org/example/SignupIntegrationTest.java (hunk 5 above)
Comment (reemessam) [reply src/test/java/org/example/SignupIntegrationTest.java:37]: for this and the others

**Comment by FearLeSS-21** [comment]
okay  working on all

________________________________

File: .gitignore [hunk 8]
```diff
@@ -9,6 +9,7 @@ target/
 .idea/compiler.xml
 .idea/libraries/
 .idea/*
+.idea
```
Comment (reemessam) [inline .gitignore:12]: you also need to remove the already commited one

File: src/main/java/org/example/validation/UserValidator.java (hunk 1 above)
Comment (FearLeSS-21) [reply src/main/java/org/example/validation/UserValidator.java:5]: must i delete the user validator  ?

Comment (reemessam) [reply src/main/java/org/example/validation/UserValidator.java:5]: you will not need it anymore if you use springboot validation

File: src/main/java/org/example/model/User.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+@AllArgsConstructor
+@Entity
+public class User {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @NotBlank(message = "Email is mandatory")
+    @Pattern(regexp = "^(?=.*example)([^@\\s]+)@([^@\\s]+\\.[^@\\s]+)$",
```
Comment (reemessam) [inline src/main/java/org/example/model/User.java:24]: i didn't notice yoou use the entity , no remove these from here and create a user dto which you will use in the controller

File: src/main/java/org/example/builder/UserBuilder.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
+package org.example.builder;
```
Comment (reemessam) [inline src/main/java/org/example/builder/UserBuilder.java:1]: no that's not what i meant ,  i meant @Builder on the class 
so for the userDto use @Builder annotation and then use it accordingly

File: src/main/java/org/example/controller/UserController.java [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+@RestController
+@RequestMapping("/users")
+@CrossOrigin(origins = "http://localhost:4200")
+public class UserController {
+
+    @Autowired
+    private UserService userService;
+
+    @PostMapping("/signup")
+    public ResponseEntity<?> signup(@Valid @RequestBody User user) {
```
Comment (reemessam) [inline src/main/java/org/example/controller/UserController.java:19]: this should be changed to take the userDto instead

File: src/main/java/org/example/service/UserService.java [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+@Service
+public class UserService {
+
+    @Autowired
+    private UserRepository userRepository;
+
+    @Autowired
+    private PasswordEncoder passwordEncoder;
+
+    public ResponseEntity<?> handleUserSignup(User user) {
```
Comment (reemessam) [inline src/main/java/org/example/service/UserService.java:21]: the service doesn't  return responseEntity the controller does
if you want to return the user , then create a user view model and return it , we can't use the user , or you can return no body with status code 201 created instead

File: src/main/java/org/example/service/UserService.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+        Optional<User> existingUser = findByEmail(user.getEmail());
+        if (existingUser.isPresent()) {
+            return ResponseEntity.status(409).body("Email already in use");
```
Comment (reemessam) [inline src/main/java/org/example/service/UserService.java:24]: create your exception and throw it , then you will need to handle in your GlobalExceptionHandler

File: src/main/java/org/example/exception/GlobalExceptionHandler.java [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+import java.util.HashMap;
+import java.util.Map;
+
+@ControllerAdvice
+public class GlobalExceptionHandler {
+
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    public ResponseEntity<Map<String, String>> handleValidationExceptions(MethodArgumentNotValidException ex) {
```
Comment (reemessam) [inline src/main/java/org/example/exception/GlobalExceptionHandler.java:16]: create a errorDto instead

File: src/main/java/org/example/dto/UserDTO.java [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+import lombok.*;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
+public class UserDTO {
+
+    @NotBlank(message = "Email is mandatory")
+    @Pattern(regexp = "^[^@\\s]+@[^@\\s]+\\.[^@\\s]+$",
```
Comment (reemessam) [inline src/main/java/org/example/dto/UserDTO.java:15]: there is @Email validator you can use

File: src/main/java/org/example/exception/UserInvalidEmailFormatException.java [hunk 16]
```diff
@@ -0,0 +1,7 @@
+package org.example.exception;
+
+public class UserInvalidEmailFormatException extends RuntimeException {
```
Comment (reemessam) [inline src/main/java/org/example/exception/UserInvalidEmailFormatException.java:3]: why did you add this , plus it's not used

File: src/main/java/org/example/exception/UserInvalidPasswordFormatException.java [hunk 17]
```diff
@@ -0,0 +1,6 @@
+package org.example.exception;
+public class UserInvalidPasswordFormatException extends RuntimeException {
```
Comment (reemessam) [inline src/main/java/org/example/exception/UserInvalidPasswordFormatException.java:2]: same

File: src/main/java/org/example/model/User.java [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    private String email;
+    private String password;
+    private String name;
+
+    public User(String email, String password, String name) {
```
Comment (reemessam) [inline src/main/java/org/example/model/User.java:23]: why did you add this, i don't think you had it before?

File: src/main/java/org/example/service/UserService.java [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+
+    public UserViewModel registerUser(UserDTO userDTO) {
+
+
+        Optional<User> existingUser = userRepository.findByEmail(userDTO.getEmail());
+        if (existingUser.isPresent()) {
+            throw new UserAlreadyExistsException("Email already in use");
+        }
+
+        User user = new User();
```
Comment (reemessam) [inline src/main/java/org/example/service/UserService.java:33]: use object mapper

File: src/main/java/org/example/service/UserService.java [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+        user.setEmail(userDTO.getEmail());
+        user.setPassword(passwordEncoder.encode(userDTO.getPassword()));
+        user.setName(userDTO.getName());
+
+        User savedUser = userRepository.save(user);
+        return UserViewModel.builder()
```
Comment (reemessam) [inline src/main/java/org/example/service/UserService.java:39]: use object mapper

File: src/test/java/org/example/SignupIntegrationTest.java [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+
+    @Autowired
+    private PasswordEncoder passwordEncoder;
+
+    @Autowired
+    private ObjectMapper objectMapper;
+
+    @BeforeEach
+    public void setup() {
+        userRepository.deleteAll();
```
Comment (reemessam) [inline src/test/java/org/example/SignupIntegrationTest.java:38]: you can also use @Transactional look it up

File: src/main/java/org/example/exception/GlobalExceptionHandler.java [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+    @ExceptionHandler(UserInvalidEmailFormatException.class)
+    public ResponseEntity<ErrorDTO> handleInvalidEmailFormat(UserInvalidEmailFormatException ex) {
+        ErrorDTO error = ErrorDTO.builder()
+                .message(ex.getMessage())
+                .details("Email format is invalid. Please provide a valid email address.")
+                .build();
+        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(error);
+    }
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    public ResponseEntity<Map<String, String>> handleValidationExceptions(MethodArgumentNotValidException ex) {
```
Comment (reemessam) [inline src/main/java/org/example/exception/GlobalExceptionHandler.java:44]: i wanted all of them to return errorDto, but you can do it in the new mr for the login 😄

File: src/main/java/org/example/model/User.java (hunk 18 above)
Comment (FearLeSS-21) [reply src/main/java/org/example/model/User.java:23]: Used constructor in testing

Comment (reemessam) [reply src/main/java/org/example/model/User.java:23]: don't need it use builder instead

File: src/main/java/org/example/config/SecurityConfig.java [hunk 23]
```diff
@@ -0,0 +1,30 @@
...
+    public PasswordEncoder passwordEncoder() {
+        return new BCryptPasswordEncoder();
+    }
+
+    @Bean
+    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+        http
+                .authorizeRequests()
+                .requestMatchers("/users/**").permitAll() // Allow access to /users endpoints
+                .anyRequest().permitAll()
```
Comment (reemessam) [inline src/main/java/org/example/config/SecurityConfig.java:25]: this should be .anyRequest().authenticated() otherwise you permit everything

File: src/main/java/org/example/service/UserService.java (hunk 20 above)
Comment (reemessam) [reply src/main/java/org/example/service/UserService.java:39]: you forgot this one

File: src/test/java/org/example/SignupIntegrationTest.java (hunk 21 above)
Comment (reemessam) [reply src/test/java/org/example/SignupIntegrationTest.java:39]: it's either not both

File: src/test/java/org/example/SignupIntegrationTest.java [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+    public void testSignupWithExistingEmail() throws Exception {
+        userRepository.save(User.builder()
+                .email("testuser@example.com")
+                .password(passwordEncoder.encode("Password123@"))
+                .name("Test User")
+                .build());
+
+        testSignup("testuser@example.com", "Password123@", "Test User", 409);
+
+        User foundUser = userRepository.findByEmail("testuser@example.com").orElse(null);
```
Comment (reemessam) [inline src/test/java/org/example/SignupIntegrationTest.java:75]: if the purpose of this assertion is to make sure that the post with exisiting email didn't change the one already there then test it with another name or password etc. otherwise both are the same so it doesn't make sense

File: src/main/java/org/example/service/UserService.java [hunk 25]
```diff
@@ -0,0 +1,56 @@
...
+    @Autowired
+    private ObjectMapper objectMapper;
+
+    public UserViewModel registerUser(UserDTO userDTO) {
+        Optional<User> existingUser = userRepository.findByEmail(userDTO.getEmail());
+        if (existingUser.isPresent()) {
+            throw new UserAlreadyExistsException("Email already in use");
+        }
+
+        User user = mapToUser(userDTO); // Changed
```
Comment (reemessam) [inline src/main/java/org/example/service/UserService.java:33]: user = objectMapper.convertValue(userDTO, User.class); directly and let the method throw the exception

File: src/main/java/org/example/service/UserService.java [hunk 26]
```diff
@@ -0,0 +1,56 @@
...
+    private User mapToUser(UserDTO userDTO) { // New method
+        try {
+            return objectMapper.convertValue(userDTO, User.class);
+        } catch (IllegalArgumentException e) {
+            throw new RuntimeException("Failed to convert UserDTO to User", e);
+        }
+    }
+
+    private UserViewModel mapToUserViewModel(User user) { // New method
+        return UserViewModel.builder()
```
Comment (reemessam) [inline src/main/java/org/example/service/UserService.java:49]: object mapper 😄
